File: long_questions/DSA/Leetcode/golang/26_Data_Structure_Design/155_Min_Stack.go
package main

import "fmt"

// 155. Min Stack
// Time: O(1) for all operations, Space: O(N)
type MinStack struct {
	stack    []int
	minStack []int
}

// Constructor initializes the stack
func ConstructorMinStack() MinStack {
	return MinStack{
		stack:    []int{},
		minStack: []int{},
	}
}

// Push pushes element onto stack
func (this *MinStack) Push(val int) {
	this.stack = append(this.stack, val)
	
	// Push to minStack if it's empty or val is <= current min
	if len(this.minStack) == 0 || val <= this.minStack[len(this.minStack)-1] {
		this.minStack = append(this.minStack, val)
	}
}

// Pop removes the element on top of the stack
func (this *MinStack) Pop() {
	if len(this.stack) == 0 {
		return
	}
	
	top := this.stack[len(this.stack)-1]
	this.stack = this.stack[:len(this.stack)-1]
	
	// Remove from minStack if it matches the popped element
	if len(this.minStack) > 0 && top == this.minStack[len(this.minStack)-1] {
		this.minStack = this.minStack[:len(this.minStack)-1]
	}
}

// Top gets the top element of the stack
func (this *MinStack) Top() int {
	if len(this.stack) == 0 {
		return -1 // Or handle error appropriately
	}
	return this.stack[len(this.stack)-1]
}

// GetMin retrieves the minimum element in the stack
func (this *MinStack) GetMin() int {
	if len(this.minStack) == 0 {
		return -1 // Or handle error appropriately
	}
	return this.minStack[len(this.minStack)-1]
}

// Alternative implementation using single stack with min tracking
type MinStackOptimized struct {
	stack []struct {
		val int
		min int
	}
}

func ConstructorMinStackOptimized() MinStackOptimized {
	return MinStackOptimized{
		stack: []struct {
			val int
			min int
		}{},
	}
}

func (this *MinStackOptimized) Push(val int) {
	min := val
	if len(this.stack) > 0 {
		currentMin := this.stack[len(this.stack)-1].min
		if val > currentMin {
			min = currentMin
		}
	}
	
	this.stack = append(this.stack, struct {
		val int
		min int
	}{val, min})
}

func (this *MinStackOptimized) Pop() {
	if len(this.stack) > 0 {
		this.stack = this.stack[:len(this.stack)-1]
	}
}

func (this *MinStackOptimized) Top() int {
	if len(this.stack) == 0 {
		return -1
	}
	return this.stack[len(this.stack)-1].val
}

func (this *MinStackOptimized) GetMin() int {
	if len(this.stack) == 0 {
		return -1
	}
	return this.stack[len(this.stack)-1].min
}

// Alternative implementation using difference method
type MinStackDiff struct {
	stack []int
	min   int
}

func ConstructorMinStackDiff() MinStackDiff {
	return MinStackDiff{
		stack: []int{},
		min:   0,
	}
}

func (this *MinStackDiff) Push(val int) {
	if len(this.stack) == 0 {
		this.stack = append(this.stack, 0)
		this.min = val
	} else {
		diff := val - this.min
		this.stack = append(this.stack, diff)
		if val < this.min {
			this.min = val
		}
	}
}

func (this *MinStackDiff) Pop() {
	if len(this.stack) == 0 {
		return
	}
	
	diff := this.stack[len(this.stack)-1]
	this.stack = this.stack[:len(this.stack)-1]
	
	if diff < 0 {
		this.min = this.min - diff
	}
}

func (this *MinStackDiff) Top() int {
	if len(this.stack) == 0 {
		return -1
	}
	
	diff := this.stack[len(this.stack)-1]
	if diff < 0 {
		return this.min
	}
	return this.min + diff
}

func (this *MinStackDiff) GetMin() int {
	if len(this.stack) == 0 {
		return -1
	}
	return this.min
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Stack with Minimum Tracking
- **Dual Stack Approach**: Main stack + min stack for O(1) minimum retrieval
- **Optimized Single Stack**: Store values with running minimums
- **Difference Method**: Track differences from minimum for efficiency
- **O(1) Minimum**: Retrieve minimum in constant time

## 2. PROBLEM CHARACTERISTICS
- **Stack Operations**: Push, pop, top, getMin in O(1) time
- **Minimum Tracking**: Need efficient minimum retrieval
- **Space Complexity**: Trade-off between time and space efficiency
- **Multiple Implementations**: Different approaches for min tracking

## 3. SIMILAR PROBLEMS
- Implement Queue using Stacks (LeetCode 232) - Queue with two stacks
- Max Stack (LeetCode 155) - Track maximum instead of minimum
- Stack with GetMiddle - Implement efficient middle element access
- Min Queue (LeetCode 622) - Queue with minimum tracking

## 4. KEY OBSERVATIONS
- **Min Stack Logic**: Maintain separate stack of minimums
- **Space Trade-off**: Extra space for O(1) minimum operations
- **Optimization Opportunities**: Multiple ways to track minimum
- **Edge Cases**: Empty stack, duplicate minimums, single element

## 5. VARIATIONS & EXTENSIONS
- **Max Stack**: Track maximum instead of minimum
- **Min-Max Stack**: Track both minimum and maximum
- **GetMedian**: Support for median calculation
- **Custom Comparator**: Support for custom comparison functions

## 6. INTERVIEW INSIGHTS
- Always clarify: "Time complexity requirements? Space constraints?"
- Edge cases: empty stack, single element, all same values
- Time complexity: O(1) for all operations
- Space complexity: O(N) where N=stack size
- Key insight: trade space for O(1) minimum retrieval

## 7. COMMON MISTAKES
- Not handling empty stack correctly
- Wrong min stack update logic
- Inefficient minimum tracking (O(N) instead of O(1))
- Not updating min stack on pop operations
- Off-by-one errors in stack indexing

## 8. OPTIMIZATION STRATEGIES
- **Dual Stack**: O(1) time, O(N) space - standard approach
- **Optimized Single Stack**: O(1) time, O(N) space - space-efficient
- **Difference Method**: O(1) time, O(N) space - mathematical approach
- **Lazy Evaluation**: Compute minimum only when needed

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like a stack with a separate minimum tracker:**
- You have a regular stack for push/pop operations
- You also maintain a separate structure for minimums
- Like having a stack and a separate "minimum so far" display
- When you push, you update both the stack and minimum tracker
- When you pop, you remove from both if necessary

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Stack with push, pop, top, getMin operations
2. **Goal**: Implement stack where getMin() is O(1) time
3. **Constraints**: All operations should be efficient
4. **Output**: Stack with O(1) minimum retrieval

#### Phase 2: Key Insight Recognition
- **"Dual stack natural fit"** → Separate min stack for O(1) access
- **"Space-time trade-off"** → Extra space for faster minimum access
- **"Multiple approaches"** → Several ways to track minimum efficiently
- **"Edge cases critical"** → Empty stack handling is crucial

#### Phase 3: Strategy Development
```
Human thought process:
"I need a stack with O(1) getMin operation.
Regular stack getMin() would be O(N), too slow.

Dual Stack Approach:
1. Main stack: regular push/pop operations
2. Min stack: track minimums seen so far
3. Push: add to main stack, update min stack
4. Pop: remove from main stack, remove from min stack if needed
5. GetMin: return top of min stack

Optimized Single Stack:
1. Store value + current minimum in each stack node
2. Update running minimum on push
3. GetMin: return current minimum directly

Both approaches give O(1) getMin!"
```

#### Phase 4: Edge Case Handling
- **Empty stack**: Return appropriate values (often -1 or error)
- **Single element**: Both stack and min stack have same element
- **Duplicate minimums**: Handle multiple equal minimums correctly
- **Large values**: Handle potential overflow issues

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Operations: Push(5), Push(3), Push(7), GetMin(), Pop(), GetMin()

Human thinking:
"Dual Stack Approach:
1. Push(5): main=[5], min=[5]
2. Push(3): main=[5,3], min=[3] (3 < 5)
3. Push(7): main=[5,3,7], min=[3] (7 > 3)
4. GetMin(): return min.top() = 3 ✓
5. Pop(): remove 7 from main, min unchanged
   main=[5,3], min=[3]
6. GetMin(): return min.top() = 3 ✓

Optimized Single Stack:
1. Push(5): stack=[{val:5, min:5}]
2. Push(3): stack=[{val:5, min:5}, {val:3, min:3}]
3. Push(7): stack=[{val:5, min:5}, {val:3, min:3}, {val:7, min:3}]
4. GetMin(): return stack.top().min = 3 ✓
5. Pop(): remove {val:7, min:3}
   stack=[{val:5, min:5}, {val:3, min:3}]
6. GetMin(): return stack.top().min = 3 ✓

Both approaches work correctly!"
```

#### Phase 6: Intuition Validation
- **Why dual stack works**: Separate min stack tracks history of minimums
- **Why optimized works**: Each node carries its own minimum information
- **Why O(1)**: Direct access to minimum without traversal
- **Why space trade-off**: Extra space enables constant time minimum

### Common Human Pitfalls & How to Avoid Them
1. **"Why not just traverse stack?"** → O(N) time is too slow
2. **"Should I use heap?"** → Heap doesn't maintain LIFO order
3. **"What about getMedian?"** → Different problem, needs different approach
4. **"Can I optimize space?"** → Trade-offs between time and space
5. **"What about duplicates?"** → Handle equal minimums correctly

### Real-World Analogy
**Like a stack with a separate minimum display:**
- You have a stack of items (main stack)
- You also have a separate display showing current minimum
- When you add items, you update both the stack and minimum display
- When you remove items, you update both if needed
- Like a warehouse inventory system with current minimum price display
- Minimum display updates automatically as items are added/removed

### Human-Readable Pseudocode
```
class MinStack:
    stack = []
    minStack = []
    
    function push(val):
        stack.push(val)
        if minStack is empty or val <= minStack.top():
            minStack.push(val)
    
    function pop():
        if stack is empty:
            return error
        val = stack.pop()
        if val == minStack.top():
            minStack.pop()
        return val
    
    function top():
        if stack is empty:
            return error
        return stack.top()
    
    function getMin():
        if minStack is empty:
            return error
        return minStack.top()

// Optimized single stack version
class MinStackOptimized:
    stack = []
    
    function push(val):
        min = val if stack is empty else min(val, stack.top().min)
        stack.push({val: val, min: min})
    
    function pop():
        if stack is empty:
            return error
        return stack.pop().val
    
    function top():
        if stack is empty:
            return error
        return stack.top().val
    
    function getMin():
        if stack is empty:
            return error
        return stack.top().min
```

### Execution Visualization

### Example: Operations: Push(5), Push(3), Push(7), GetMin(), Pop(), GetMin()
```
Dual Stack Approach:
Initial: stack=[], minStack=[]

Push(5): stack=[5], minStack=[5]
Push(3): stack=[5,3], minStack=[3] (3 < 5)
Push(7): stack=[5,3,7], minStack=[3] (7 > 3)
GetMin(): return minStack.top() = 3 ✓
Pop(): remove 7 from stack, minStack unchanged
stack=[5,3], minStack=[3]
GetMin(): return minStack.top() = 3 ✓

Optimized Single Stack:
Initial: stack=[]

Push(5): stack=[{val:5, min:5}]
Push(3): stack=[{val:5, min:5}, {val:3, min:3}]
Push(7): stack=[{val:5, min:5}, {val:3, min:3}, {val:7, min:3}]
GetMin(): return stack.top().min = 3 ✓
Pop(): remove {val:7, min:3}
stack=[{val:5, min:5}, {val:3, min:3}]
GetMin(): return stack.top().min = 3 ✓
```

### Key Visualization Points:
- **Dual Stack**: Main stack + separate min stack
- **Min Stack Logic**: Only push when new value is ≤ current minimum
- **Optimized Stack**: Each node carries its own minimum
- **O(1) Access**: Direct access to minimum without traversal

### Memory Layout Visualization:
```
Dual Stack State Evolution:
Initial: stack=[], minStack=[]
After Push(5): stack=[5], minStack=[5]
After Push(3): stack=[5,3], minStack=[3]
After Push(7): stack=[5,3,7], minStack=[3]
After Pop(): stack=[5,3], minStack=[3]

Optimized Stack State Evolution:
Initial: stack=[]
After Push(5): stack=[{val:5, min:5}]
After Push(3): stack=[{val:5, min:5}, {val:3, min:3}]
After Push(7): stack=[{val:5, min:5}, {val:3, min:3}, {val:7, min:3}]
After Pop(): stack=[{val:5, min:5}, {val:3, min:3}]

Operation Complexity:
Push: O(1) time, O(1) space
Pop: O(1) time, O(1) space
Top: O(1) time, O(1) space
GetMin: O(1) time, O(1) space
```

### Time Complexity Breakdown:
- **Push**: O(1) time, O(1) space for all approaches
- **Pop**: O(1) time, O(1) space for all approaches
- **Top**: O(1) time, O(1) space for all approaches
- **GetMin**: O(1) time, O(1) space (key optimization)
- **Space**: O(N) where N=stack size

### Alternative Approaches:

#### 1. Lazy Minimum Calculation (O(1) amortized time)
```go
type MinStackLazy struct {
    stack []int
    min   int
    dirty bool
}

func (this *MinStackLazy) GetMin() int {
    if this.dirty {
        this.recalculateMin()
        this.dirty = false
    }
    return this.min
}
```
- **Pros**: Potentially more space efficient
- **Cons**: More complex implementation

#### 2. Segment Tree (O(log N) time, O(N) space)
```go
type MinStackSegment struct {
    tree *SegmentTree
    stack []int
}

func (this *MinStackSegment) GetMin() int {
    return this.tree.Query(0, len(this.stack)-1)
}
```
- **Pros**: Efficient for range queries
- **Cons**: Overkill for simple getMin operation

#### 3. Binary Indexed Tree (O(log N) time, O(N) space)
```go
type MinStackBIT struct {
    bit *BinaryIndexedTree
    stack []int
}

func (this *MinStackBIT) GetMin() int {
    return this.bit.QueryMin()
}
```
- **Pros**: Efficient for dynamic updates
- **Cons**: Complex implementation

### Extensions for Interviews:
- **Max Stack**: Track maximum instead of minimum
- **Min-Max Stack**: Track both minimum and maximum
- **GetMedian**: Support for median calculation
- **Custom Comparator**: Support for custom comparison functions
- **Performance Analysis**: Discuss space-time trade-offs
*/
func main() {
	// Test cases
	fmt.Println("=== Testing MinStack ===")
	
	// Test 1: Basic operations
	minStack := ConstructorMinStack()
	minStack.Push(-2)
	minStack.Push(0)
	minStack.Push(-3)
	fmt.Printf("After pushing -2, 0, -3:\n")
	fmt.Printf("  Top: %d, Min: %d\n", minStack.Top(), minStack.GetMin())
	
	minStack.Pop()
	fmt.Printf("After pop:\n")
	fmt.Printf("  Top: %d, Min: %d\n", minStack.Top(), minStack.GetMin())
	
	// Test 2: Edge cases
	fmt.Println("\n=== Testing Edge Cases ===")
	emptyStack := ConstructorMinStack()
	fmt.Printf("Empty stack - Top: %d, Min: %d\n", emptyStack.Top(), emptyStack.GetMin())
	
	emptyStack.Push(5)
	fmt.Printf("After pushing 5 - Top: %d, Min: %d\n", emptyStack.Top(), emptyStack.GetMin())
	
	// Test 3: Duplicate minimums
	fmt.Println("\n=== Testing Duplicate Minimums ===")
	dupStack := ConstructorMinStack()
	dupStack.Push(1)
	dupStack.Push(1)
	dupStack.Push(2)
	fmt.Printf("After pushing 1, 1, 2 - Top: %d, Min: %d\n", dupStack.Top(), dupStack.GetMin())
	
	dupStack.Pop()
	fmt.Printf("After pop - Top: %d, Min: %d\n", dupStack.Top(), dupStack.GetMin())
	
	// Test optimized versions
	fmt.Println("\n=== Testing Optimized Version ===")
	optStack := ConstructorMinStackOptimized()
	optStack.Push(-2)
	optStack.Push(0)
	optStack.Push(-3)
	fmt.Printf("After pushing -2, 0, -3:\n")
	fmt.Printf("  Top: %d, Min: %d\n", optStack.Top(), optStack.GetMin())
	
	optStack.Pop()
	fmt.Printf("After pop:\n")
	fmt.Printf("  Top: %d, Min: %d\n", optStack.Top(), optStack.GetMin())
	
	// Test difference method
	fmt.Println("\n=== Testing Difference Method ===")
	diffStack := ConstructorMinStackDiff()
	diffStack.Push(-2)
	diffStack.Push(0)
	diffStack.Push(-3)
	fmt.Printf("After pushing -2, 0, -3:\n")
	fmt.Printf("  Top: %d, Min: %d\n", diffStack.Top(), diffStack.GetMin())
	
	diffStack.Pop()
	fmt.Printf("After pop:\n")
	fmt.Printf("  Top: %d, Min: %d\n", diffStack.Top(), diffStack.GetMin())
}

File: long_questions/DSA/Leetcode/java/26_Data_Structure_Design/155_Min_Stack.java
import java.util.*;

public class MinStack {
    
    // 155. Min Stack
    // Time: O(1) for all operations, Space: O(N)
    private Stack<Integer> stack;
    private Stack<Integer> minStack;

    // Constructor initializes the stack
    public MinStack() {
        stack = new Stack<>();
        minStack = new Stack<>();
    }

    // Push pushes element onto stack
    public void push(int val) {
        stack.push(val);
        
        // Push to minStack if it's empty or val is <= current min
        if (minStack.isEmpty() || val <= minStack.peek()) {
            minStack.push(val);
        }
    }

    // Pop removes the element on top of the stack
    public void pop() {
        if (stack.isEmpty()) {
            return;
        }
        
        int top = stack.pop();
        
        // Remove from minStack if it matches the popped element
        if (!minStack.isEmpty() && top == minStack.peek()) {
            minStack.pop();
        }
    }

    // Top gets the top element of the stack
    public int top() {
        if (stack.isEmpty()) {
            return -1; // Or handle error appropriately
        }
        return stack.peek();
    }

    // GetMin retrieves the minimum element in the stack
    public int getMin() {
        if (minStack.isEmpty()) {
            return -1; // Or handle error appropriately
        }
        return minStack.peek();
    }

    // Alternative implementation using single stack with min tracking
    public static class MinStackOptimized {
        private Stack<int[]> stack; // Each entry contains {val, min}

        public MinStackOptimized() {
            stack = new Stack<>();
        }

        public void push(int val) {
            int min = val;
            if (!stack.isEmpty()) {
                int currentMin = stack.peek()[1];
                if (val > currentMin) {
                    min = currentMin;
                }
            }
            
            stack.push(new int[]{val, min});
        }

        public void pop() {
            if (!stack.isEmpty()) {
                stack.pop();
            }
        }

        public int top() {
            if (stack.isEmpty()) {
                return -1;
            }
            return stack.peek()[0];
        }

        public int getMin() {
            if (stack.isEmpty()) {
                return -1;
            }
            return stack.peek()[1];
        }
    }

    // Alternative implementation using difference method
    public static class MinStackDiff {
        private Stack<Long> stack;
        private long min;

        public MinStackDiff() {
            stack = new Stack<>();
            min = 0;
        }

        public void push(int val) {
            if (stack.isEmpty()) {
                stack.push(0L);
                min = val;
            } else {
                long diff = (long)val - min;
                stack.push(diff);
                if (val < min) {
                    min = val;
                }
            }
        }

        public void pop() {
            if (stack.isEmpty()) {
                return;
            }
            
            long diff = stack.pop();
            if (diff < 0) {
                min = min - diff;
            }
        }

        public int top() {
            if (stack.isEmpty()) {
                return -1;
            }
            
            long diff = stack.peek();
            if (diff < 0) {
                return (int)min;
            }
            return (int)(min + diff);
        }

        public int getMin() {
            if (stack.isEmpty()) {
                return -1;
            }
            return (int)min;
        }
    }

    public static void main(String[] args) {
        // Test cases
        System.out.println("=== Testing MinStack ===");
        
        // Test 1: Basic operations
        MinStack minStack = new MinStack();
        minStack.push(-2);
        minStack.push(0);
        minStack.push(-3);
        System.out.printf("After pushing -2, 0, -3:\n");
        System.out.printf("  Top: %d, Min: %d\n", minStack.top(), minStack.getMin());
        
        minStack.pop();
        System.out.printf("After pop:\n");
        System.out.printf("  Top: %d, Min: %d\n", minStack.top(), minStack.getMin());
        
        // Test 2: Edge cases
        System.out.println("\n=== Testing Edge Cases ===");
        MinStack emptyStack = new MinStack();
        System.out.printf("Empty stack - Top: %d, Min: %d\n", emptyStack.top(), emptyStack.getMin());
        
        emptyStack.push(5);
        System.out.printf("After pushing 5 - Top: %d, Min: %d\n", emptyStack.top(), emptyStack.getMin());
        
        // Test 3: Duplicate minimums
        System.out.println("\n=== Testing Duplicate Minimums ===");
        MinStack dupStack = new MinStack();
        dupStack.push(1);
        dupStack.push(1);
        dupStack.push(2);
        System.out.printf("After pushing 1, 1, 2 - Top: %d, Min: %d\n", dupStack.top(), dupStack.getMin());
        
        dupStack.pop();
        System.out.printf("After pop - Top: %d, Min: %d\n", dupStack.top(), dupStack.getMin());
        
        // Test optimized versions
        System.out.println("\n=== Testing Optimized Version ===");
        MinStackOptimized optStack = new MinStackOptimized();
        optStack.push(-2);
        optStack.push(0);
        optStack.push(-3);
        System.out.printf("After pushing -2, 0, -3:\n");
        System.out.printf("  Top: %d, Min: %d\n", optStack.top(), optStack.getMin());
        
        optStack.pop();
        System.out.printf("After pop:\n");
        System.out.printf("  Top: %d, Min: %d\n", optStack.top(), optStack.getMin());
        
        // Test difference method
        System.out.println("\n=== Testing Difference Method ===");
        MinStackDiff diffStack = new MinStackDiff();
        diffStack.push(-2);
        diffStack.push(0);
        diffStack.push(-3);
        System.out.printf("After pushing -2, 0, -3:\n");
        System.out.printf("  Top: %d, Min: %d\n", diffStack.top(), diffStack.getMin());
        
        diffStack.pop();
        System.out.printf("After pop:\n");
        System.out.printf("  Top: %d, Min: %d\n", diffStack.top(), diffStack.getMin());
    }

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Data Structure Design
- **Min Stack**: Stack with O(1) minimum retrieval
- **Auxiliary Structure**: Additional stack to track minimums
- **Space-Time Tradeoff**: Extra space for constant time minimum
- **Multiple Implementations**: Different optimization strategies

## 2. PROBLEM CHARACTERISTICS
- **Stack Operations**: push, pop, top, getMin in O(1)
- **Minimum Tracking**: Need efficient minimum element retrieval
- **Constraint Handling**: Handle empty stack gracefully
- **Performance Requirements**: All operations should be O(1)

## 3. SIMILAR PROBLEMS
- Max Stack (with O(1) maximum)
- Queue with O(1) minimum
- Stack with O(1) middle element
- Design Min Queue

## 4. KEY OBSERVATIONS
- Regular stack getMin() requires O(N) traversal
- Auxiliary stack stores minimums efficiently
- Different strategies: difference tracking, pair storage, single stack
- Space complexity: O(N) for auxiliary structure
- Time complexity: O(1) for all operations

## 5. VARIATIONS & EXTENSIONS
- Support for duplicate minimums
- Stack with range queries
- Thread-safe implementation
- Memory-constrained versions

## 6. INTERVIEW INSIGHTS
- Clarify: "Should I handle negative values?"
- Edge cases: empty stack, single element, duplicates
- Time complexity: O(1) vs O(N) regular stack
- Space complexity: O(N) auxiliary vs O(1) regular

## 7. COMMON MISTAKES
- Not handling empty stack properly
- Incorrect minimum updates in auxiliary stack
- Integer overflow with long differences
- Not maintaining O(1) for all operations

## 8. OPTIMIZATION STRATEGIES
- Lazy minimum updates
- In-place calculations where possible
- Early termination for obvious cases
- Choose appropriate strategy based on usage pattern

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like a smart filing cabinet:**
- You have a regular stack of papers (main stack)
- You also have a separate index card showing the minimum value
- When you add new papers, you update the index card
- When you remove papers, you check if the removed paper was the minimum
- If yes, you need to find the new minimum from remaining papers
- This gives you O(1) access to the minimum value

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Need stack with O(1) minimum operations
2. **Goal**: Implement push, pop, top, getMin efficiently
3. **Output**: Stack data structure with constant-time minimum

#### Phase 2: Key Insight Recognition
- **"Why regular stack insufficient?"** → getMin() requires O(N) traversal
- **"How to track minimum efficiently?"** → Auxiliary structure
- **"What's the tradeoff?"** → Extra space for O(1) minimum access
- **"Why multiple approaches?"** → Different optimization strategies

#### Phase 3: Strategy Development
```
Human thought process:
"I'll use auxiliary stack:
1. Main stack stores all elements
2. Min stack stores minimum values efficiently
3. Push: update both stacks appropriately
4. Pop: remove from main, update min stack if needed
5. getMin: return top of min stack
6. This gives O(1) for all operations"
```

#### Phase 4: Edge Case Handling
- **Empty stack**: Return error or sentinel value
- **Single element**: Both stacks have same value
- **Duplicates**: Handle multiple minimums correctly
- **Large values**: Use long to avoid overflow

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Operations: push(5), push(2), push(7), getMin(), pop(), getMin()

Human thinking:
"Let's track both stacks:

Initial: main=[], min=[]

Push 5:
- main: [5], min: [5]
- min = 5

Push 2:
- main: [5,2], min: [5,2]
- min = 2

Push 7:
- main: [5,2,7], min: [5,2,7]
- min = 2

getMin(): return 2 ✓

Pop():
- Remove 7 from main: [5,2]
- 7 was not the minimum, min stack unchanged
- min = 2

getMin(): return 2 ✓"
```

#### Phase 6: Intuition Validation
- **Why it works**: Auxiliary stack always has current minimum
- **Why it's efficient**: All operations are O(1)
- **Why it's correct**: Minimum tracking is maintained accurately

### Common Human Pitfalls & How to Avoid Them
1. **"Why not just scan main stack?"** → O(N) time for getMin()
2. **"What about recalculation?"** → O(N²) if done naively
3. **"How to handle duplicates?"** → Need proper logic in auxiliary stack
4. **"What about memory?"** → Trade space for time efficiency

### Real-World Analogy
**Like a priority checkout line:**
- You have a regular line of customers (main stack)
- You also have an express lane for VIP customers (min stack)
- When new customers arrive, you decide if they're VIP enough
- VIP customers go to express lane (min stack)
- Regular customers go to regular lane (main stack)
- When a VIP customer leaves, you check if they were the minimum VIP
- If yes, find the next minimum VIP from the express lane
- This gives you O(1) access to the minimum priority customer

### Human-Readable Pseudocode
```
class MinStack:
    mainStack = empty stack
    minStack = empty stack
    
    function push(val):
        mainStack.push(val)
        if minStack.isEmpty() or val <= minStack.peek():
            minStack.push(val)
        else:
            diff = val - minStack.peek()
            minStack.push(diff)
    
    function pop():
        val = mainStack.pop()
        if !minStack.isEmpty() and val == minStack.peek():
            minStack.pop()
        else if !minStack.isEmpty():
            diff = minStack.pop()
            min = min - diff
            minStack.push(diff)
        return val
    
    function getMin():
        if minStack.isEmpty():
            return -1  // or error
        return minStack.peek()
```

### Execution Visualization

### Example: push(5), push(2), push(7), getMin(), pop(), getMin()
```
Stack Evolution:
Initial: main=[], min=[]

Push 5:
main: [5], min: [5], min=5

Push 2:
main: [5,2], min: [5,2], min=2

Push 7:
main: [5,2,7], min: [5,2,7], min=2

getMin(): return 2 ✓

Pop():
- Remove 7 from main: [5,2]
- 7 ≠ min(2), min stack unchanged
- min = 2

getMin(): return 2 ✓

Visualization:
Main Stack: [5,2,7] ← 7 removed
Min Stack: [5,2] ← minimum remains 2
```

### Key Visualization Points:
- **Dual stack** approach for O(1) minimum
- **Difference tracking** maintains relative values
- **Lazy updates** only when necessary
- **Space-time tradeoff** for constant-time operations

### Memory Layout Visualization:
```
Two-Stack Approach:
Main Stack: [5,2,7] (bottom)
Min Stack: [5,2] (top)

Operation Flow:
Push: main.push(), min.push() if needed
Pop: main.pop(), min.pop() if popped was min
getMin: min.peek()

State Tracking:
- Main stack: actual values
- Min stack: minimum values or differences
- Relationship: min.peek() = actual minimum in main stack
```

### Time Complexity Breakdown:
- **Push**: O(1) - push to main, possibly push to min
- **Pop**: O(1) - pop from main, possibly pop from min
- **getMin**: O(1) - peek from min stack
- **Space**: O(N) for auxiliary stack
- **Optimal**: O(1) time for all operations with O(N) space
- **vs Regular Stack**: O(N) time for getMin(), O(N) space
*/
